File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverse(self,head,k,group):
        temp=head
        prev=None
        cnt=0

        while temp!=None and cnt<k:
            forward=temp.next
            temp.next=prev
            prev=temp
            temp=forward
            cnt=cnt+1
        if forward!=None and group!=0:
            head.next=self.reverse(forward,k,group-1)
        else:
            head.next=forward
        return prev
    def length(self,head):
        cnt=0
        temp=head
        while temp!=None:
            cnt=cnt+1
            temp=temp.next
        return cnt
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head==None or head.next==None:
            return head
        size=self.length(head)
        groups=size//k
        return self.reverse(head,k,groups-1)
        

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int groups=0;
    public int lengthOfLL(ListNode head){
        int length=0;
        ListNode temp=head;
        while(temp!=null){
            length=length+1;
            temp=temp.next;
        }
        return length;
    }
    public ListNode counter(ListNode head,int k){
        int count=1;
        ListNode temp=head;
        while(temp!=null){
            if(count==k){
                break;
            }
            count++;
            temp=temp.next;
        }
        return temp;
    }
    public ListNode reverse(ListNode head){
        if(head==null || head.next==null){
            return head;
        }
        ListNode prev=null;
        ListNode temp=null;
        while(head!=null){
            temp=head.next;
            head.next=prev;
            prev=head;
            head=temp;
        }
        return prev;
    }
    public ListNode helper(ListNode head,int times,int k){
        if(head==null|| head.next==null || times>groups){
            return head;
        }
        ListNode reversePoint=counter(head,k);
        ListNode newReversedHead=helper(reversePoint.next,times+1,k);
        reversePoint.next=null;
        ListNode reversedHead=reverse(head);
        head.next=newReversedHead;
        return reversedHead;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        int length=lengthOfLL(head);
        groups=(length/k);
        return helper(head,1,k);
    }
}